Add support for the Aho-Corasick algorithm in regexes.

teo-tsirpanis opened the pull request:
Fixes #62447.

To efficiently skip through text in search of a match, the .NET regex engine employs a number of shortcuts, such as searching for a leading string, a leading set, specific characters in a fixed distance from the start, or relying on anchors to stick to the input's beginning or end. These optimizations are contained in the `RegexFindOptimizations` class and enumerated in the `FindNextStartingPositionMode` type.

This PR introduces a new optimization mode named `LeadingMultiString_LeftToRight`. It searches for multiple strings in one pass by walking through the nodes of a trie using the Aho-Corasick algorithm. This trie is constructed by walking the regex's `RegexNode` from left to right, taking into account the many possible branches it can take through sets and alternations, adding nodes to the trie during the process, and stopping if it sees an unrecognized pattern. For example the pattern `a(b+|c|d)?e` will create a trie that matches the words `ab`, `ace`, `ade` and `ae`.

The algorithm itself has a slight difference from the standard Aho-Corasick. Because we don't care which of the words in the trie it found, instead of the "dictionary link" transitions, each node has a "match length" property, which holds the length of the word that matches in this node, saving some extra indirections to get it. Another difference is in the matching logic, where extra care is taken to return the leftmost match. The code responsible for the tree link construction and the actual string matching resides in the internal `MultiStringMatcher` class.

As explained before, the way the algorithm is integrated with the rest of the engine is through the `RegexFindOptimizations` class. Because this class is used during matching only in the interpreted mode, for the compiled mode I added a field of type `MultiStringMatcher` in the `CompiledRegexRunner` class, which is accessed when needed. For the source-generated regex engine, we emit a trie node type and matching algorithm as a shared utility, and for each regex engine we emit an array of trie nodes which we feed to the matching algorithm.

~~An outstanding question is __when to use Aho-Corasick__. According to https://github.com/dotnet/runtime/issues/69682#issuecomment-1171195266 vectorized searching methods should be preferred over AC. I could have put it in the end of the `RegexFindOptimizations`' constructor, using it if only other methods fail but it feels too weak to me; I fear it would be sidestepped in favor of methods that might turn out to be worse, with my suspicions falling to `LeadingString` (for example `aaa(b|c|d|e|f)` would not use AC and use a `LeadingString` of `aaa`).~~

For this reason I made it try to build a trie in the very beginning and only after the optimizations for regexes with anchors, and if the trie can be made and has at least __5__ matches[^1] it will be ~~used~~ accepted, otherwise it tries to find a single leading prefix, and if that does not exist as well, it continues with the other methods. The trie's construction is skipped entirely if we are in the non-backtracking mode (seems excessive since we will already use a DFA), the compiled mode (on similar pattern sizes it is slower with AC), or we are matching from right to left (I wonder what it would take to support this). __New:__ In later commits the logic was refactored. If the trie is preliminarily accepted, it will not be actually used unless the other vectorized modes have been tried and none can be used.

I have also placed some limits on the trie generation. A trie cannot have more than __128__ nodes[^2] -or __256__ in the compiled mode or the source generator-, and a word in a trie cannot be longer than __12__ characters. If the former limit is reached, no new trie nodes will be added during the regex node's traversal, leading to prematurely marching states as matching, which has a chance of rejecting the whole trie (for example if we have an alternation of 100 patterns, the trie gets full in the first 50 and at least one of the other 50 needs adding a node, the trie will not be able to produce accurate results). If the latter limit is reached, word will be truncated.

All tests succeed on my machine. I have ran benchmarks similar to https://github.com/dotnet/runtime/issues/62447#issuecomment-1134002629, comparing the existing and modified regex engines, measuring both the creation of `Regex` objects and counting matches on them, and the results are available in [benchmarks.zip](https://github.com/dotnet/runtime/files/9033786/benchmarks.zip). In short, performance significantly increased when Aho-Corasick is being used, especially in the interpreted engine, but creating `Regex` objects regressed. There are two sets of results, both with and without the aforementioned trie node limits.

[^1]: Based on benchmarks on https://github.com/dotnet/runtime/issues/69682#issuecomment-1171173565
[^2]: A trie with the 50 most common English words has 91 nodes.

teo-tsirpanis committed:
Add MultiStringMatcher.

teo-tsirpanis committed:
Refactor MultiStringMatcher, move TrieNode to a new file and remove WordId.

teo-tsirpanis committed:
Support building tries from RegexNodes.

teo-tsirpanis committed:
Rename Trie to TrieBuilder and directly use the List<TrieNode> in matching.

teo-tsirpanis committed:
Add the new code to the source generator.

teo-tsirpanis committed:
Do not allow the root node to be a match.

teo-tsirpanis committed:
Allow constructing a MultiStringMatcher from a trie and add some utilities.

teo-tsirpanis committed:
Integrate MultiStringSearcher in the interpreted regex engine via RegexFindOptimizations.

teo-tsirpanis committed:
Fix some bugs and add the new files in the unit test project.

teo-tsirpanis committed:
Fix CreatePrefixMatcher's logic.

teo-tsirpanis committed:
Support MultiStringMatcher in the compiled engine.

teo-tsirpanis committed:
Remove `TrieExtensions.GetCommonPrefix` and always use `FindPrefix`.

It does not return what we want. In "(ab|ab*)bc" for example, the trie correctly matches a and abbc. The trie's common prefix is abbc, but a match does not have to start with it. GetPrefix returns "a" which is correct but not optimal, it could have been "ab", but the algorithm would need to be made much more complicated to detect that.

teo-tsirpanis committed:
Support MultiStringMatcher in the Regex source generator.

teo-tsirpanis committed:
Replace TrieNode.DictionaryLink with MatchLength.

It is simpler and easier to explain, and we can use it since we don't care which of the words we matched.

teo-tsirpanis committed:
Remove the unused MultiStringMatcher constructor that accepts a span of strings.

teo-tsirpanis committed:
Limit the maximum depth of a trie node.

teo-tsirpanis committed:
Allow the root node to be a match, and discard the trie if it becomes.

teo-tsirpanis committed:
Avoid an allocation during Dictionary enumeration.

teo-tsirpanis committed:
Fix the branching depth logic.

Do not let branching in alternation node's children affect each other.

teo-tsirpanis committed:
Avoid adding duplicate items in NodeCollections.

teo-tsirpanis committed:
Replace the branching depth limit with a simpler node count limit.

And add a debug-only property on TrieNode to easily see the children count.

teo-tsirpanis committed:
Fix a build error in Release mode.

teo-tsirpanis committed:
Make the source-generated trie node children representation more compact.

Nodes with no children will be assigned a cached empty dictionary to reduce allocations, and nodes with one child will be declared in one line.

teo-tsirpanis committed:
Replace a call with callvirt in the compiler.

teo-tsirpanis committed:
Rename TrieBuilder.TryCreateFromPrefix and make it return a nullable list of nodes.

teo-tsirpanis committed:
Refactor the trie builder's handling of loops and add support for "x?" patterns.

teo-tsirpanis committed:
Pass the TrieBuilder around by reference.
The type is a struct. It makes no difference now but could subtly break the code in the future.

teo-tsirpanis committed:
Assert that each trie node with no children is a match node.

AaronRobinsonMSFT commented:
/cc @stephentoub

teo-tsirpanis commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/TrieBuilder.cs:
```diff
@@ -0,0 +1,509 @@
+#endif
+                };
+
+                node.Children.Add(c, nextNodeIndex);
```
I had the idea that if the regex is case-insensitive, it would add a transition via all case-equivalent characters to the same new node, but I need a `CutureInfo` to use `RegexCaseEquivalences`, and it seems to me that at this point, the case-insensitivity of the regex has been stripped away. 🤔

stephentoub replied:
There's no difference between "(?i)ab" and "[Aa][Bb]".  If the goal is to apply a transition for all case-equivalent characters, you'd just do so for all members of the set.

teo-tsirpanis replied:
An idea I had is to merge equivalent states after constructing the trie links. It could greatly reduce the state count on case-insensitive regexes, but it needs some thought on how to do, would happen after taking the trie node limit into account, and might be expensive. I'd prefer to explore this in the future.

stephentoub commented:
Thanks. I'll be offline for a few days; I'll review later this week.

teo-tsirpanis committed:
Store the trie in an array of structs when matching.

Also refactor the trie link creation code and always store a trie node's path when needed.
Benchmarks have shown some gains.

teo-tsirpanis commented:
I ran the benchmarks with smaller word sizes. For those unfamiliar, the benchmarks count the first `WordCount` most common English words in [a random Project Gutenberg book](https://www.gutenberg.org/ebooks/68148). After 50 words we hit the node limit of 128 and Aho-Corasick is not used.

|             Method | WordCount |  Options |        Mean |     Error |    StdDev | Ratio | RatioSD | Rank |
|------------------- |---------- |--------- |------------:|----------:|----------:|------:|--------:|-----:|
|    **CountWordsRegex** |        **10** |     **None** |  **4,476.0 μs** | **113.32 μs** | **334.12 μs** |  **1.00** |    **0.00** |    **2** |
| CountWordsRegexNeo |        10 |     None |  3,554.9 μs |  66.44 μs | 140.13 μs |  0.79 |    0.06 |    1 |
|                    |           |          |             |           |           |       |         |      |
|    **CountWordsRegex** |        **10** | **Compiled** |    **907.3 μs** |  **18.09 μs** |  **42.98 μs** |  **1.00** |    **0.00** |    **1** |
| CountWordsRegexNeo |        10 | Compiled |  2,591.2 μs |  51.34 μs | 133.44 μs |  2.86 |    0.20 |    2 |
|                    |           |          |             |           |           |       |         |      |
|    **CountWordsRegex** |        **25** |     **None** |  **7,303.7 μs** | **145.60 μs** | **391.15 μs** |  **1.00** |    **0.00** |    **2** |
| CountWordsRegexNeo |        25 |     None |  3,922.4 μs |  70.03 μs |  54.67 μs |  0.55 |    0.02 |    1 |
|                    |           |          |             |           |           |       |         |      |
|    **CountWordsRegex** |        **25** | **Compiled** |  **1,163.1 μs** |  **21.67 μs** |  **20.27 μs** |  **1.00** |    **0.00** |    **1** |
| CountWordsRegexNeo |        25 | Compiled |  2,636.1 μs |  52.57 μs |  53.99 μs |  2.27 |    0.07 |    2 |
|                    |           |          |             |           |           |       |         |      |
|    **CountWordsRegex** |        **50** |     **None** | **14,321.8 μs** | **282.73 μs** | **423.17 μs** |  **1.00** |    **0.00** |    **2** |
| CountWordsRegexNeo |        50 |     None |  5,168.8 μs | 101.73 μs | 227.54 μs |  0.36 |    0.02 |    1 |
|                    |           |          |             |           |           |       |         |      |
|    **CountWordsRegex** |        **50** | **Compiled** |  **2,336.4 μs** |  **46.32 μs** |  **69.33 μs** |  **1.00** |    **0.00** |    **1** |
| CountWordsRegexNeo |        50 | Compiled |  3,554.2 μs |  68.02 μs |  90.81 μs |  1.52 |    0.07 |    2 |

At these sizes the `Compiled` engine does not get much help from Aho-Corasick. I disabled it there.

---

I then refactored the trie node representation in the `MultiStringMatcher` class. The nodes are value types and stored in an array, reducing indiection. The benchmarks were positive:

|             Method | WordCount | Options |      Mean |     Error |    StdDev | Ratio | RatioSD | Rank |
|------------------- |---------- |-------- |----------:|----------:|----------:|------:|--------:|-----:|
|    **CountWordsRegex** |        **10** |    **None** |  **4.298 ms** | **0.1152 ms** | **0.3325 ms** |  **1.00** |    **0.00** |    **2** |
| CountWordsRegexNeo |        10 |    None |  3.140 ms | 0.0627 ms | 0.0939 ms |  0.72 |    0.06 |    1 |
|                    |           |         |           |           |           |       |         |      |
|    **CountWordsRegex** |        **25** |    **None** |  **7.373 ms** | **0.1467 ms** | **0.3220 ms** |  **1.00** |    **0.00** |    **2** |
| CountWordsRegexNeo |        25 |    None |  3.722 ms | 0.0675 ms | 0.0599 ms |  0.51 |    0.03 |    1 |
|                    |           |         |           |           |           |       |         |      |
|    **CountWordsRegex** |        **50** |    **None** | **14.421 ms** | **0.2645 ms** | **0.2474 ms** |  **1.00** |    **0.00** |    **2** |
| CountWordsRegexNeo |        50 |    None |  4.723 ms | 0.0938 ms | 0.0921 ms |  0.33 |    0.01 |    1 |

teo-tsirpanis committed:
Avoid expanding the trie past match nodes.

Since we don't care about the length of the match, when we see patterns like "the|their|them|they", we can only add "the" to the trie.

teo-tsirpanis committed:
Return from the Aho-Corasick search immediately upon finding a match.

teo-tsirpanis committed:
Do not use MultiStringMatcher on some more problematic tries.

teo-tsirpanis commented:
After the last changes that took advantage that we return the leftmost match, and can immediately exit if we found one, performance increased a bit more:

|             Method | WordCount | Options |      Mean |     Error |    StdDev | Ratio | Rank |
|------------------- |---------- |-------- |----------:|----------:|----------:|------:|-----:|
|    **CountWordsRegex** |        **10** |    **None** |  **4.039 ms** | **0.0606 ms** | **0.0506 ms** |  **1.00** |    **2** |
| CountWordsRegexNeo |        10 |    None |  2.720 ms | 0.0360 ms | 0.0319 ms |  0.67 |    1 |
|                    |           |         |           |           |           |       |      |
|    **CountWordsRegex** |        **25** |    **None** |  **6.928 ms** | **0.0761 ms** | **0.0635 ms** |  **1.00** |    **2** |
| CountWordsRegexNeo |        25 |    None |  3.186 ms | 0.0278 ms | 0.0217 ms |  0.46 |    1 |
|                    |           |         |           |           |           |       |      |
|    **CountWordsRegex** |        **50** |    **None** | **14.352 ms** | **0.0903 ms** | **0.0801 ms** |  **1.00** |    **2** |
| CountWordsRegexNeo |        50 |    None |  3.845 ms | 0.0544 ms | 0.0482 ms |  0.27 |    1 |

teo-tsirpanis committed:
Fix two typos.

teo-tsirpanis committed:
Test that some patterns get optimized to LeadingMultiString_LeftToRight.

teo-tsirpanis commented:
I refactored the logic around selecting whether to use AC, and we prefer the vectorized modes over it. In the benchmarks on a `WordCount` of 10 it uses `LeadingSet_LeftToRight`. But for some unexplained reason it's still quite faster; I couldn't believe it, I ran the benchmark many times, nothing else changed. 🤔

|             Method | WordCount | Options |      Mean |     Error |    StdDev |    Median | Ratio | RatioSD | Rank |
|------------------- |---------- |-------- |----------:|----------:|----------:|----------:|------:|--------:|-----:|
|    **CountWordsRegex** |        **10** |    **None** |  **4.253 ms** | **0.0848 ms** | **0.1843 ms** |  **4.190 ms** |  **1.00** |    **0.00** |    **2** |
| CountWordsRegexNeo |        10 |    None |  3.282 ms | 0.0618 ms | 0.1650 ms |  3.213 ms |  0.78 |    0.05 |    1 |
|                    |           |         |           |           |           |           |       |         |      |
|    **CountWordsRegex** |        **25** |    **None** |  **7.145 ms** | **0.0995 ms** | **0.0777 ms** |  **7.123 ms** |  **1.00** |    **0.00** |    **2** |
| CountWordsRegexNeo |        25 |    None |  3.276 ms | 0.0206 ms | 0.0172 ms |  3.280 ms |  0.46 |    0.00 |    1 |
|                    |           |         |           |           |           |           |       |         |      |
|    **CountWordsRegex** |        **50** |    **None** | **15.248 ms** | **0.2930 ms** | **0.6117 ms** | **15.066 ms** |  **1.00** |    **0.00** |    **2** |
| CountWordsRegexNeo |        50 |    None |  4.208 ms | 0.0850 ms | 0.2383 ms |  4.202 ms |  0.28 |    0.02 |    1 |

teo-tsirpanis commented:
@stephentoub I just learned that new feature work for 7.0 will complete on Tuesday. If you can review this soon I will do my best to respond to your feedback quickly enough to reach the deadline.

stephentoub commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -335,6 +382,65 @@ private static void AddIsECMABoundaryHelper(Dictionary<string, string[]> require
+                    "    }",
+                    "",
+                    "    return -1;",
+                    "}"
```
This approach still requires the compiled code to effectively interpret the trie, reading each character, looking up the next node, etc.  An alternative is to actually compile the trie logic down into code, emitting custom code for each state, with each state having its own label that can be jumped to.  Did you consider something like that?  Imagine instead of `if (trie[currentState].Children.TryGetValue(c, out int nextState))`, you have code for the dedicated `currentState` that contains a `switch (c)`, with a case for everything in the trie, each hardcoded with a goto for the right next state, etc. It would be optimized for the exact trie in use, it would avoid the need for the boilerplate helpers, etc., and the built-in limitations on the number of nodes in the trie should limit code explosion.

teo-tsirpanis replied:
Yes, I thought of this but left it for another time. I will give it a try.

teo-tsirpanis replied:
I did it for `Compiled` regexes and will do it in the source generator soon.

teo-tsirpanis replied:
Done. I will benchmark it tomorrow.

stephentoub commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -335,6 +382,65 @@ private static void AddIsECMABoundaryHelper(Dictionary<string, string[]> require
+                    "                currentState = nextState;",
+                    "                break;",
+                    "            }",
+                    "            if (currentState == 0)",
```
```suggestion
                    "",
                    "            if (currentState == 0)",
```

stephentoub commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -744,6 +854,19 @@ void EmitIndexOf_LeftToRight()
                 }
             }
 
+            void EmitMultiStringSearch()
```
```suggestion
            // Emits a trie-based search for one of multiple possible prefixes that can start a match.
            void EmitMultiStringSearch()
```

stephentoub commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -4420,9 +4543,9 @@ private static string MatchCharacterClass(RegexOptions options, string chExpr, s
-                        (true,  true)  => $"((((ch = {chExpr}) | 0x{mask:X}) != {Literal((char)(setChars[1] | mask))}) & (ch != {Literal(setChars[2])}))",
+                        (true, false) => $"(((ch = {chExpr}) != {Literal(setChars[0])}) & (ch != {Literal(setChars[1])}) & (ch != {Literal(setChars[2])}))",
+                        (false, true) => $"((((ch = {chExpr}) | 0x{mask:X}) == {Literal((char)(setChars[1] | mask))}) | (ch == {Literal(setChars[2])}))",
+                        (true, true) => $"((((ch = {chExpr}) | 0x{mask:X}) != {Literal((char)(setChars[1] | mask))}) & (ch != {Literal(setChars[2])}))",
```
Please revert these changes.  They were lined up on the arrow.

stephentoub commented on src/libraries/System.Text.RegularExpressions/gen/Stubs.cs:
```diff
@@ -25,7 +22,7 @@ public static unsafe StringBuilder Append(this StringBuilder stringBuilder, Read
         public static ReadOnlyMemory<char>[] GetChunks(this StringBuilder stringBuilder)
         {
-            var chars = new char[stringBuilder.Length];
+            char[] chars = new char[stringBuilder.Length];
```
Revert

teo-tsirpanis replied:
Reverted.

stephentoub commented on src/libraries/System.Text.RegularExpressions/gen/Stubs.cs:
```diff
@@ -38,6 +35,10 @@ internal static class StringExtensions
+            if (length == 0)
+            {
+                return string.Empty;
+            }
```
Where is this being used with length 0 and where does that optimization matter?

teo-tsirpanis replied:
Reverted.

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/CompiledRegexRunner.cs:
```diff
@@ -19,10 +24,11 @@ internal sealed class CompiledRegexRunner : RegexRunner
         internal delegate void ScanDelegate(RegexRunner runner, ReadOnlySpan<char> text);
 
-        public CompiledRegexRunner(ScanDelegate scan, CultureInfo? culture)
+        public CompiledRegexRunner(ScanDelegate scan, CultureInfo? culture, MultiStringMatcher? prefixMatcher)
```
My comments about the source generator emitting code specific to a trie apply here as well.

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/MultiStringMatcher.cs:
```diff
@@ -0,0 +1,167 @@
+    /// </summary>
+    internal sealed class MultiStringMatcher
+    {
+        private readonly TrieNodeWithLinks[] _trie;
+
+        internal ReadOnlySpan<TrieNodeWithLinks> Trie => _trie;
+
```
```suggestion
```

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/MultiStringMatcher.cs:
```diff
@@ -0,0 +1,167 @@
+        /// </summary>
+        internal MultiStringMatcher(List<TrieNode> trie)
+        {
+            _trie = BuildTrieLinks(trie);
+        }
+
```
```suggestion
            Trie = BuildTrieLinks(trie);
        }

        internal ReadOnlySpan<TrieNodeWithLinks> Trie { get; }
```

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/MultiStringMatcher.cs:
```diff
@@ -0,0 +1,167 @@
+            TrieNodeWithLinks[] result = new TrieNodeWithLinks[trie.Count];
+            Queue<int> vertexQueue = new Queue<int>();
+            vertexQueue.Enqueue(TrieNode.Root);
+#if REGEXGENERATOR
+            while (vertexQueue.Count > 0)
+            {
+                int currentVertex = vertexQueue.Dequeue();
+#else
+            while (vertexQueue.TryDequeue(out int currentVertex))
+            {
+#endif
```
```suggestion
            while (vertexQueue.Count > 0)
            {
                int currentVertex = vertexQueue.Dequeue();
```

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/MultiStringMatcher.cs:
```diff
@@ -0,0 +1,167 @@
+
+        private static TrieNodeWithLinks[] BuildTrieLinks(List<TrieNode> trie)
+        {
+            TrieNodeWithLinks[] result = new TrieNodeWithLinks[trie.Count];
+            Queue<int> vertexQueue = new Queue<int>();
+            vertexQueue.Enqueue(TrieNode.Root);
```
```suggestion
            TrieNodeWithLinks[] result = new TrieNodeWithLinks[trie.Count];

            Queue<int> vertexQueue = new Queue<int>();
            vertexQueue.Enqueue(TrieNode.Root);

```

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/MultiStringMatcher.cs:
```diff
@@ -0,0 +1,167 @@
+                int suffixLink;
+                int matchLength;
+
+                if (node.IsRoot)
+                {
+                    suffixLink = TrieNode.Root;
+                    matchLength = -1;
+                    goto End;
+                }
+
+                // one character substrings
+                if (node.Parent == TrieNode.Root)
+                {
+                    suffixLink = TrieNode.Root;
+                    matchLength = node.IsMatch ? node.Depth : result[suffixLink].MatchLength;
+                    goto End;
+                }
+
```
This doesn't need to use gotos... it can just be:
```C#
if (node.IsRoot)
{
    suffixLink = TrieNode.Root;
    matchLength = -1
}
else if (node.Parent == TrieNode.Root)
{
    suffixLink = TrieNode.Root;
    matchLength = node.IsMatch ? node.Depth : result[suffixLink].MatchLength;
}
else
{
    // To calculate the suffix link for the current vertex, we need the suffix
    ...
}

result[currentVertex] = new TrieNodeWithLinks()
...
```

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/MultiStringMatcher.cs:
```diff
@@ -0,0 +1,167 @@
+                    {
+                        break;
+                    }
+                    // Jump by suffix links until we reach the root or find a better prefix for the current substring.
```
```suggestion

                    // Jump by suffix links until we reach the root or find a better prefix for the current substring.
```

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/MultiStringMatcher.cs:
```diff
@@ -0,0 +1,167 @@
+                        suffixLink = TrieNode.Root;
+                        break;
+                    }
+                    // Go up by suffixlink
```
```suggestion

                    // Go up by SuffixLink
```

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/MultiStringMatcher.cs:
```diff
@@ -0,0 +1,167 @@
+                        currentState = nextState;
+                        break;
+                    }
+                    if (currentState == TrieNode.Root)
```
```suggestion

                    if (currentState == TrieNode.Root)
```

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexPrefixAnalyzer.cs:
```diff
@@ -12,8 +11,89 @@ namespace System.Text.RegularExpressions
+        /// </summary>
+        internal static (string? SinglePrefix, List<TrieNode>? PrefixTrie) CreatePrefixMatcher(RegexNode node)
+        {
+            if ((node.Options & (RegexOptions.RightToLeft | RegexOptions.NonBacktracking | RegexOptions.Compiled)) != 0)
```
I'm confused now.  Why do the source generator and RegexCompiler have code in support of the prefix matcher if we disable this for RegexOptions.Compiled?

teo-tsirpanis replied:
I added this relatively late after seeing disappointing results for the compiled mode. I should fine-tune it more and re-enable it. Perhaps directly emitting the AC logic would improve things.

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFindOptimizations.cs:
```diff
@@ -179,6 +182,12 @@ public RegexFindOptimizations(RegexNode root, RegexOptions options)
                         _asciiLookups = new uint[fixedDistanceSets.Count][];
+                        if (FindMode == FindNextStartingPositionMode.FixedDistanceSets_LeftToRight || fixedDistanceSets[0].Chars is null)
+                        {
+                            // FixedDistanceSets and LeadingSet with null Chars are not vectorized.
```
I don't understand this condition.  It's jumping to PreferLeadingMultiStringIfAvailable if FindMode == FindNextStartingPositionMode.FixedDistanceSets_LeftToRight, regardless of whether fixedDistanceSets[0].Chars is null, but FixedDistanceSets is absolutely vectorizable when Chars is not null.

teo-tsirpanis replied:
Answered in https://github.com/dotnet/runtime/pull/71588#discussion_r916656081.

teo-tsirpanis replied:
I removed the first part of the `||`. We now only check if `Chars` is null.

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFindOptimizations.cs:
```diff
@@ -724,6 +768,9 @@ internal enum FindNextStartingPositionMode
         /// <summary>An "endz" anchor at the end of the pattern, with the pattern always matching a fixed-length expression.</summary>
         TrailingAnchor_FixedLength_LeftToRight_EndZ,
 
+        // The following modes have vectorized implementations.
+        // TODO: LeadingSet is not always vectorized. Move the non-vectorized case to a new LeadingCharClass mode.
+
```
```suggestion
```

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFindOptimizations.cs:
```diff
@@ -742,6 +789,12 @@ internal enum FindNextStartingPositionMode
         /// <summary>A multi-character case-sensitive string at a fixed distance from the start of the pattern.</summary>
         FixedDistanceString_LeftToRight,
 
+        // The following modes are not vectorized.
```
This is not true.

teo-tsirpanis replied:
I should have said "not _fully_ vectorized". I see some `IndexOf(Any)?` calls on these modes, but they still loop over each character and call `RegexCharClass.CharInClass` which can be pretty heavy if the character is not ASCII or not cached.

Is my understanding correct that we should prefer `LeadingMultiString` over modes that call `CharInClass`?

teo-tsirpanis replied:
I was refering to https://github.com/dotnet/runtime/blob/d344e72a84b98d84b3e34d37e9c59cf04930cf5b/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFindOptimizations.cs#L601-L604 but upon closer look this is technically an $O(1)$ operation since we call `CharInClass` once per set. In any case I fixed the vectorized mode detection logic.

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/TrieBuilder.cs:
```diff
@@ -0,0 +1,555 @@
+                    return new NodeCollection(newNode);
+                default:
+                    canContinue = false;
+#if REGEXGENERATOR
+                    HashSet<int> visited = new HashSet<int>();
+#else
+                    HashSet<int> visited = new HashSet<int>(count);
+#endif
```
```suggestion
                    HashSet<int> visited = new HashSet<int>(
#if !REGEXGENERATOR
                       count
#endif
                    );
```

teo-tsirpanis replied:
It fails with `error SA1112: (NETCORE_ENGINEERING_TELEMETRY=Build) Closing parenthesis should be on line of opening parenthesis`. I will just create the hash set without the initial size.

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/TrieBuilder.cs:
```diff
@@ -0,0 +1,555 @@
+                        {
+                            continue;
+                        }
+                        // See comments in handling RegexNodeKind.Alternate below, to understand how it works.
```
```suggestion

                        // See comments in handling RegexNodeKind.Alternate below, to understand how it works.
```

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/TrieBuilder.cs:
```diff
@@ -0,0 +1,555 @@
+                            _nodes[newNode].IsMatch = true;
+                        }
+                    }
+                    return new NodeCollection(result);
```
```suggestion

                    return new NodeCollection(result);
```

stephentoub commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/TrieBuilder.cs:
```diff
@@ -0,0 +1,555 @@
+                goto End;
+            }
+
+            while (regexNode.Kind is RegexNodeKind.Capture or RegexNodeKind.Group or RegexNodeKind.Atomic)
```
```suggestion
            while (regexNode.Kind is RegexNodeKind.Capture or RegexNodeKind.Atomic)
```

stephentoub reviewed (changes requested):
Thanks for working on this.

Can you point me to the latest benchmark code you're running?

teo-tsirpanis committed:
Address most PR feedback.

teo-tsirpanis committed:
Use MultiStringMatcher only if no vectorized find optimization mode can be used.

teo-tsirpanis commented:
> Can you point me to the latest benchmark code you're running?

I run the benchmarks by copying the library's source into a separate project, making the minimal changes to make it compile (convert a `switch`e over a `ReadOnlySpan<char>` into a sequence of `SequenceEquals` and disable the `SR` stuff), and renaming `Regex` and `RegexOptions` to `RegexNeo` and `RegexNeoOptions`. I will upload a complete benchmark project once I finish with directly emitting the AC logic in the compiled mode.

teo-tsirpanis committed:
Emit specialized Aho-Corasick logic in the compiled mode instead of using MultiStringMatcher.

teo-tsirpanis committed:
Remove unreachable trie nodes.

teo-tsirpanis commented:
Thanks for the advice, I will make use of it once I finish with the remaining coding. Here are the benchmarks I will use:

[RegexBenchmark.zip](https://github.com/dotnet/runtime/files/9074303/RegexBenchmark.zip)

teo-tsirpanis committed:
Do not use MultiStringMatcher in the non-backtracking engine.

teo-tsirpanis committed:
Emit specialized Aho-Corasick logic in the source generator instead of traversing an array of trie nodes.

teo-tsirpanis commented:
Here are the results:

### CountBenchmarks

|     Method |        Job |         Toolchain | WordCount |  Options |        Mean |    Error |    StdDev |      Median | Ratio | RatioSD |
|----------- |----------- |------------------ |---------- |--------- |------------:|---------:|----------:|------------:|------:|--------:|
| **CountWords** | **Job-HXWVTR** | **\main\corerun.exe** |        **10** |     **None** |  **3,100.1 μs** | **58.63 μs** |  **51.97 μs** |  **3,094.5 μs** |  **1.00** |    **0.00** |
| CountWords | Job-YGVHGA |   \pr\corerun.exe |        10 |     None |  3,016.6 μs | 59.44 μs | 137.76 μs |  2,953.1 μs |  1.00 |    0.04 |
|            |            |                   |           |          |             |          |           |             |       |         |
| **CountWords** | **Job-HXWVTR** | **\main\corerun.exe** |        **10** | **Compiled** |    **773.9 μs** | **12.33 μs** |  **11.54 μs** |    **781.0 μs** |  **1.00** |    **0.00** |
| CountWords | Job-YGVHGA |   \pr\corerun.exe |        10 | Compiled |    782.6 μs | 14.28 μs |  13.36 μs |    789.4 μs |  1.01 |    0.03 |
|            |            |                   |           |          |             |          |           |             |       |         |
| **CountWords** | **Job-HXWVTR** | **\main\corerun.exe** |        **25** |     **None** |  **5,497.8 μs** | **87.56 μs** |  **73.11 μs** |  **5,531.4 μs** |  **1.00** |    **0.00** |
| CountWords | Job-YGVHGA |   \pr\corerun.exe |        25 |     None |  3,187.3 μs | 63.61 μs | 135.56 μs |  3,162.7 μs |  0.60 |    0.02 |
|            |            |                   |           |          |             |          |           |             |       |         |
| **CountWords** | **Job-HXWVTR** | **\main\corerun.exe** |        **25** | **Compiled** |  **1,119.5 μs** | **20.35 μs** |  **18.04 μs** |  **1,126.3 μs** |  **1.00** |    **0.00** |
| CountWords | Job-YGVHGA |   \pr\corerun.exe |        25 | Compiled |    883.5 μs | 13.41 μs |  12.54 μs |    892.3 μs |  0.79 |    0.01 |
|            |            |                   |           |          |             |          |           |             |       |         |
| **CountWords** | **Job-HXWVTR** | **\main\corerun.exe** |        **50** |     **None** | **11,819.0 μs** | **58.51 μs** |  **51.87 μs** | **11,822.8 μs** |  **1.00** |    **0.00** |
| CountWords | Job-YGVHGA |   \pr\corerun.exe |        50 |     None |  3,710.1 μs | 22.54 μs |  18.83 μs |  3,708.0 μs |  0.31 |    0.00 |
|            |            |                   |           |          |             |          |           |             |       |         |
| **CountWords** | **Job-HXWVTR** | **\main\corerun.exe** |        **50** | **Compiled** |  **1,570.9 μs** | **24.43 μs** |  **22.85 μs** |  **1,579.8 μs** |  **1.00** |    **0.00** |
| CountWords | Job-YGVHGA |   \pr\corerun.exe |        50 | Compiled |  1,213.9 μs | 18.20 μs |  17.02 μs |  1,214.6 μs |  0.77 |    0.02 |

### CreateBenchmarks

|      Method |        Job |         Toolchain | WordCount |      Mean |     Error |    StdDev | Ratio | RatioSD |   Gen 0 | Allocated |
|------------ |----------- |------------------ |---------- |----------:|----------:|----------:|------:|--------:|--------:|----------:|
| **CreateRegex** | **Job-HXWVTR** | **\main\corerun.exe** |        **10** |  **6.317 μs** | **0.0439 μs** | **0.0410 μs** |  **1.00** |    **0.00** |  **1.9379** |      **6 KB** |
| CreateRegex | Job-YGVHGA |   \pr\corerun.exe |        10 |  9.740 μs | 0.0826 μs | 0.0732 μs |  1.54 |    0.01 |  4.1199 |     13 KB |
|             |            |                   |           |           |           |           |       |         |         |           |
| **CreateRegex** | **Job-HXWVTR** | **\main\corerun.exe** |        **25** | **14.080 μs** | **0.2747 μs** | **0.2294 μs** |  **1.00** |    **0.00** |  **4.0283** |     **12 KB** |
| CreateRegex | Job-YGVHGA |   \pr\corerun.exe |        25 | 22.766 μs | 0.3151 μs | 0.2947 μs |  1.62 |    0.03 |  9.0027 |     28 KB |
|             |            |                   |           |           |           |           |       |         |         |           |
| **CreateRegex** | **Job-HXWVTR** | **\main\corerun.exe** |        **50** | **30.095 μs** | **0.5642 μs** | **0.5278 μs** |  **1.00** |    **0.00** |  **7.9041** |     **24 KB** |
| CreateRegex | Job-YGVHGA |   \pr\corerun.exe |        50 | 53.245 μs | 0.3273 μs | 0.3061 μs |  1.77 |    0.03 | 18.3105 |     56 KB |

teo-tsirpanis committed:
Fix an outdated comment.

teo-tsirpanis committed:
Vectorize MultiStringMatcher if the root node has at most five children.

teo-tsirpanis committed:
Correctly detect when FixedDistanceSets is vectorized, and give more priority to vectorized LeadingMultiString.

teo-tsirpanis commented:
If the root of the trie has at most five children, the searching algorithm now performs a vectorized `IndexOfAny` to quickly jump to the next occurence of them. Also in this case `LeadingMultiString` is given more priority over other modes.

Aho-Corasick now makes a difference with an alternation of 10 words:

|     Method |        Job |         Toolchain | WordCount |  Options |        Mean |     Error |    StdDev |      Median | Ratio | RatioSD |
|----------- |----------- |------------------ |---------- |--------- |------------:|----------:|----------:|------------:|------:|--------:|
| **CountWords** | **Job-XYYDHH** | **\main\corerun.exe** |        **10** |     **None** |  **3,076.6 μs** |  **66.86 μs** | **192.91 μs** |  **2,970.7 μs** |  **1.00** |    **0.00** |
| CountWords | Job-FNCNJP |   \pr\corerun.exe |        10 |     None |  2,119.1 μs |  30.36 μs |  26.91 μs |  2,129.1 μs |  0.64 |    0.03 |
|            |            |                   |           |          |             |           |           |             |       |         |
| **CountWords** | **Job-XYYDHH** | **\main\corerun.exe** |        **10** | **Compiled** |    **770.6 μs** |  **12.76 μs** |  **11.94 μs** |    **764.4 μs** |  **1.00** |    **0.00** |
| CountWords | Job-FNCNJP |   \pr\corerun.exe |        10 | Compiled |    765.4 μs |   7.58 μs |   7.09 μs |    768.2 μs |  0.99 |    0.02 |
|            |            |                   |           |          |             |           |           |             |       |         |
| **CountWords** | **Job-XYYDHH** | **\main\corerun.exe** |        **25** |     **None** |  **5,405.6 μs** | **106.22 μs** | **104.32 μs** |  **5,346.4 μs** |  **1.00** |    **0.00** |
| CountWords | Job-FNCNJP |   \pr\corerun.exe |        25 |     None |  3,135.8 μs |  61.30 μs |  51.19 μs |  3,144.2 μs |  0.58 |    0.01 |
|            |            |                   |           |          |             |           |           |             |       |         |
| **CountWords** | **Job-XYYDHH** | **\main\corerun.exe** |        **25** | **Compiled** |  **1,136.8 μs** |  **15.11 μs** |  **14.13 μs** |  **1,141.8 μs** |  **1.00** |    **0.00** |
| CountWords | Job-FNCNJP |   \pr\corerun.exe |        25 | Compiled |    881.0 μs |  13.09 μs |  12.24 μs |    876.1 μs |  0.78 |    0.01 |
|            |            |                   |           |          |             |           |           |             |       |         |
| **CountWords** | **Job-XYYDHH** | **\main\corerun.exe** |        **50** |     **None** | **11,919.2 μs** | **216.99 μs** | **181.20 μs** | **11,942.5 μs** |  **1.00** |    **0.00** |
| CountWords | Job-FNCNJP |   \pr\corerun.exe |        50 |     None |  3,970.2 μs |  78.87 μs | 153.82 μs |  4,043.4 μs |  0.34 |    0.01 |
|            |            |                   |           |          |             |           |           |             |       |         |
| **CountWords** | **Job-XYYDHH** | **\main\corerun.exe** |        **50** | **Compiled** |  **1,605.9 μs** |  **31.25 μs** |  **29.24 μs** |  **1,609.4 μs** |  **1.00** |    **0.00** |
| CountWords | Job-FNCNJP |   \pr\corerun.exe |        50 | Compiled |  1,211.6 μs |  17.38 μs |  16.26 μs |  1,206.3 μs |  0.75 |    0.02 |

I will soon check word counts of less than ten, and see if lowering the match limit of five would be beneficial.

teo-tsirpanis committed:
Lower the minimum match count in the interpreted mode.

Benchmarks have shown it's profitable.

teo-tsirpanis commented:
I ran the benchmark with smaller word counts and the results were satisfactory.

|     Method |        Job |         Toolchain | WordCount |  Options |       Mean |       Error |    StdDev | Ratio | RatioSD |
|----------- |----------- |------------------ |---------- |--------- |-----------:|------------:|----------:|------:|--------:|
| **CountWords** | **Job-SODFAG** | **\main\corerun.exe** |         **2** |     **None** |   **995.1 μs** |   **868.21 μs** |  **47.59 μs** |  **1.00** |    **0.00** |
| CountWords | Job-WWAKPB |   \pr\corerun.exe |         2 |     None |   670.7 μs |   236.26 μs |  12.95 μs |  0.67 |    0.03 |
|            |            |                   |           |          |            |             |           |       |         |
| **CountWords** | **Job-SODFAG** | **\main\corerun.exe** |         **2** | **Compiled** |   **181.4 μs** |    **36.97 μs** |   **2.03 μs** |  **1.00** |    **0.00** |
| CountWords | Job-WWAKPB |   \pr\corerun.exe |         2 | Compiled |   248.1 μs |    34.42 μs |   1.89 μs |  1.37 |    0.03 |
|            |            |                   |           |          |            |             |           |       |         |
| **CountWords** | **Job-SODFAG** | **\main\corerun.exe** |         **3** |     **None** | **1,225.2 μs** | **1,193.82 μs** |  **65.44 μs** |  **1.00** |    **0.00** |
| CountWords | Job-WWAKPB |   \pr\corerun.exe |         3 |     None |   797.4 μs |   630.72 μs |  34.57 μs |  0.65 |    0.03 |
|            |            |                   |           |          |            |             |           |       |         |
| **CountWords** | **Job-SODFAG** | **\main\corerun.exe** |         **3** | **Compiled** |   **288.4 μs** |   **117.19 μs** |   **6.42 μs** |  **1.00** |    **0.00** |
| CountWords | Job-WWAKPB |   \pr\corerun.exe |         3 | Compiled |   261.5 μs |   119.35 μs |   6.54 μs |  0.91 |    0.01 |
|            |            |                   |           |          |            |             |           |       |         |
| **CountWords** | **Job-SODFAG** | **\main\corerun.exe** |         **5** |     **None** | **2,075.1 μs** | **2,072.61 μs** | **113.61 μs** |  **1.00** |    **0.00** |
| CountWords | Job-WWAKPB |   \pr\corerun.exe |         5 |     None | 1,652.8 μs |   952.66 μs |  52.22 μs |  0.80 |    0.07 |
|            |            |                   |           |          |            |             |           |       |         |
| **CountWords** | **Job-SODFAG** | **\main\corerun.exe** |         **5** | **Compiled** |   **609.5 μs** |   **773.25 μs** |  **42.38 μs** |  **1.00** |    **0.00** |
| CountWords | Job-WWAKPB |   \pr\corerun.exe |         5 | Compiled |   532.8 μs |    50.28 μs |   2.76 μs |  0.88 |    0.07 |

In the interpreted mode minimum match nodes a trie has to have was lowered to 2. For the compiled mode and the source generator the limit stays to 5, out of abundance of caution.

teo-tsirpanis commented:
And by this time all PR feedback has been addressed.

teo-tsirpanis commented:
Any news @stephentoub?

stephentoub commented:
I haven't had a chance yet to re-review in depth and try out a bunch of different workloads with it.

stephentoub commented:
Thanks for working on this, @teo-tsirpanis.

I just pulled down the change and ran some of our perf tests locally.  There are some monstrous regressions and a few nice improvements.  These are the ones I ran that were > 20% change in either direction:

Regressions:

| Type                                      | Toolchain             | Pattern                                     | Options                  | Mean                    | Ratio               |
| ----------------------------------------- | --------------------- | ------------------------------------------- | ------------------------ | ----------------------: | ------------------: |
| RustLang_Sherlock | main              | (?i)Sher[a-z]+\|Hol[a-z]+               | Compiled             | 750,554.43 ns       | 1.00            |
| RustLang_Sherlock     | pr                    | (?i)Sher[a-z]+\|Hol[a-z]+                   | Compiled                 | 1,021,485.00 ns         | 1.36                |
|                                           |                       |                                             |                          |                         |                     |
| RustLang_Sherlock | main              | (?i)Holmes                              | None                 | 594,383.47 ns       | 1.00            |
| RustLang_Sherlock     | pr                    | (?i)Holmes                                  | None                     | 1,312,564.02 ns         | 2.21                |
|                                           |                       |                                             |                          |                         |                     |
| RustLang_Sherlock | main              | (?i)Sherlock                            | None                 | 137,081.49 ns       | 1.00            |
| RustLang_Sherlock     | pr                    | (?i)Sherlock                                | None                     | 1,102,775.38 ns         | 8.14                |
|                                           |                       |                                             |                          |                         |                     |
| RustLang_Sherlock | main              | (?i)Sherlock                            | Compiled             | 107,513.19 ns       | 1.00            |
| RustLang_Sherlock     | pr                    | (?i)Sherlock                                | Compiled                 | 440,607.52 ns           | 4.12                |
|                                           |                       |                                             |                          |                         |                     |
| RustLang_Sherlock | main              | (?i)Sherlock Holmes                     | None                 | 138,625.11 ns       | 1.00            |
| RustLang_Sherlock     | pr                    | (?i)Sherlock Holmes                         | None                     | 1,070,227.53 ns         | 7.64                |
|                                           |                       |                                             |                          |                         |                     |
| RustLang_Sherlock | main              | (?i)Sherlock Holmes                     | Compiled             | 104,803.62 ns       | 1.00            |
| RustLang_Sherlock     | pr                    | (?i)Sherlock Holmes                         | Compiled                 | 437,145.39 ns           | 4.22                |
|                                           |                       |                                             |                          |                         |                     |
| RustLang_Sherlock | main              | (?i)the                                | None                 | 1,292,574.06 ns     | 1.00            |
| RustLang_Sherlock     | pr                    | (?i)the                                     | None                     | 2,081,759.17 ns         | 1.60                |
|                                           |                       |                                             |                          |                         |                     |
| SliceSlice        | main              | ?                                       | IgnoreCase           | 2,440,860,550.00 ns | 1.00            |
| SliceSlice            | pr                    | ?                                           | IgnoreCase               | 7,523,069,250.00 ns     | 3.08                |
|                                           |                       |                                             |                          |                         |                     |
| SliceSlice        | main              | ?                                       | IgnoreCase, Compiled | 1,761,960,781.82 ns | 1.00            |
| SliceSlice            | pr                    | ?                                           | IgnoreCase, Compiled     | 2,793,406,900.00 ns     | 1.56                |
|                                           |                       |                                             |                          |                         |                     |
| RustLang_Sherlock | main              | Sherlock\|Holm(...)er\|John\|Baker[45]  | None                 | 2,620,646.63 ns     | 1.00            |
| RustLang_Sherlock     | pr                    | Sherlock\|Holm(...)er\|John\|Baker[45]      | None                     | 8,500,348.72 ns         | 3.24                |
|                                           |                       |                                             |                          |                         |                     |
| Leipzig           | main              | (?i)Twain                               | None                 | 7,782,317.92 ns     | 1.00            |
| Leipzig               | pr                    | (?i)Twain                                   | None                     | 33,096,301.19 ns        | 4.25                |
|                                           |                       |                                             |                          |                         |                     |
| Leipzig           | main              | (?i)Twain                               | Compiled             | 5,961,453.75 ns     | 1.00            |
| Leipzig               | pr                    | (?i)Twain                                   | Compiled                 | 13,382,420.24 ns        | 2.25                |

Improvements:

| Type                                      | Toolchain             | Pattern                                     | Options                  | Mean                    | Ratio                |
| ----------------------------------------- | --------------------- | ------------------------------------------- | ------------------------ | ----------------------: | -------------------: |
| RustLang_Sherlock | main              | (?i)Sher[a-z]+\|Hol[a-z]+               | None                 | 5,326,859.81 ns     | 1.00             |
| RustLang_Sherlock     | pr                    | (?i)Sher[a-z]+\|Hol[a-z]+                   | None                     | 2,443,060.83 ns         | 0.46                 |
|                                           |                       |                                             |                          |                         |                      |
| RustLang_Sherlock | main              | Sher[a-z]+\|Hol[a-z]+                   | None                 | 235,633.83 ns       | 1.00             |
| RustLang_Sherlock     | pr                    | Sher[a-z]+\|Hol[a-z]+                       | None                     | 186,024.60 ns           | 0.77                 |
|                                           |                       |                                             |                          |                         |                      |
| Leipzig           | main              | Huck[a-zA-Z]+\|Saw[a-zA-Z]+             | None                 | 5,500,710.00 ns     | 1.00             |
| Leipzig               | pr                    | Huck[a-zA-Z]+\|Saw[a-zA-Z]+                 | None                     | 3,634,715.16 ns         | 0.66                 |
|                                           |                       |                                             |                          |                         |                      |
| Leipzig           | main              | Tom.{10,25}river\|river.{10,25}Tom      | None                 | 56,661,678.57 ns    | 1.00             |
| Leipzig               | pr                    | Tom.{10,25}river\|river.{10,25}Tom          | None                     | 27,020,938.46 ns        | 0.48                 |
|                                           |                       |                                             |                          |                         |                      |
| Leipzig           | main              | Tom\|Sawyer\|Huckleberry\|Finn          | None                 | 15,556,603.29 ns    | 1.00             |
| Leipzig               | pr                    | Tom\|Sawyer\|Huckleberry\|Finn              | None                     | 7,779,389.24 ns         | 0.50                 |

This might just come down to more fine-tuning of the heuristics used to decide when to employ which searching strategy, but at present I don't feel comfortable trying to squeeze this in for .NET 7.  I would rather take our time, get it "right", and get it in early for .NET 8.  Just to set expectations.

danmoseley commented:
Making no merge to keep it out of .NET 7 per comment above. Thanks for work on this @teo-tsirpanis

teo-tsirpanis commented:
All right, I will work on it later when I find time. 7 already has a boatload of improvements either way.

danmoseley commented:
The improvements seen above are really nice. Definitely seems worth continuing to work on the heuristics.

teo-tsirpanis committed:
Fix some comments.

danmoseley commented:
I'm going to close this, for clarity, until you have time to have another look @teo-tsirpanis . this does'nt mean we don't want it - just it's on the timeline of your choice. Feel free to reopen!

teo-tsirpanis commented:
Reopening.

teo-tsirpanis committed:
Preliminarily block regexes without an alternation in their fixed leading part from being optimized with LeadingMultiString.

teo-tsirpanis committed:
Eliminate the unreachable nodes during the trie link assignment.

It is more efficient (the List<TrieNode> is not created twice) and became a bit more powerful.

teo-tsirpanis committed:
Use a cached empty trie children dictionary.

teo-tsirpanis committed:
Fix compile errors.

teo-tsirpanis commented:
### Status update

* I modified the heuristics so that patterns whose leading fixed part does not contain an alternation are not eligible for `LeadingMultiString`. Regressions on case-insensitive patterns like `(?i)Sherlock Holmes` were fixed.
* The unreachable node elimination was changed to take place later, at the time of assigning suffix link and match length values, allowing some more eliminations. For example if we have pattern `bc|abcd`, the trie will match only words `bc` and `abc`.
* Nodes without children will use a cached empty `Dictionary<char, int>` instance to reduce memory footprint.

### What else needs to be done

* [ ] Fix test failures.
* [ ] Do not favor `LeadingMultiString` if ~~all~~ most its nodes' suffix links point to the root, even if the search for the first character is vectorized. This happens with patterns like `(?i)Sher[a-z]+|Hol[a-z]+`.
    * We need to do this only in compiled mode; this underutilization of Aho-Corasick in the interpreted mode shows improvements for a very strange reason.
* [ ] (maybe) [Collapse equivalent trie nodes.](https://github.com/dotnet/runtime/pull/71588#discussion_r917255988) We could do this with a variation of the [Myhill-Nerode theorem](https://www.geeksforgeeks.org/minimization-of-dfa-using-myhill-nerode-theorem/) or a lighter algorithm if I can think of one (MN is for DFAs, we have tries).
    * A problem with it is the interaction with the match node limit. If a trie has five words of length A and five words of a different length B, after collapsing the states, it will have only two match nodes. If we are in compiled mode do we keep it?

teo-tsirpanis committed:
Fix and simplify the unreachable node removal code.

teo-tsirpanis committed:
Move some test patterns without alternations back to FixedDistanceChars.

teo-tsirpanis committed:
Fix detection of alternations inside loops.

teo-tsirpanis commented:
@stephentoub as you can see from the message above I fixed the regression on case-insensitive patterns without branches, and I have ran out of ideas. The second item of the todo list will not make it accept `Tom|Sawyer|Huckleberry|Finn` and reject `Sherlock|Holm(…)er|John|Baker[45]`, and I have ran out of ideas.

I will reopen and undraft, and will appreciate your insights once you have time.

teo-tsirpanis committed:
Merge branch 'main' into aho-corasick

buyaa-n commented:
>I will reopen and undraft, and will appreciate your insights once you have time.

Sorry for delayed feedback @teo-tsirpanis, now many people are on vacation this holiday season and most likely you will not get any feedback until January.  As the PR is not ready for review, I am going to put it back to draft.

> as you can see from the message above I fixed the regression on case-insensitive patterns without branches, and I have ran out of ideas. The second item of the todo list will not make it accept Tom|Sawyer|Huckleberry|Finn and reject Sherlock|Holm(…)er|John|Baker[45], and I have ran out of ideas.

Meanwhile if you could share benchmark results showing the fix progress could be useful for @stephentoub to evaluate the progress and derive ideas to suggest

danmoseley commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFindOptimizations.cs:
```diff
@@ -76,6 +78,18 @@ public RegexFindOptimizations(RegexNode root, RegexOptions options)
             }
 
+            // If the trie's root has at most five children, we can perform a vectorized search to find them.
+            // In this case we prefer LeadingMultiString over other potentially vecrotized modes, because this
```
```suggestion
            // In this case we prefer LeadingMultiString over other potentially vectorized modes, because this
```

teo-tsirpanis replied:
Thanks!

teo-tsirpanis committed:
Update src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFindOptimizations.cs

Co-authored-by: Dan Moseley <danmose@microsoft.com>

teo-tsirpanis committed:
Merge branch 'main' into aho-corasick

teo-tsirpanis committed:
Adjust preprocessor directives.

teo-tsirpanis committed:
Adjust tests.

teo-tsirpanis committed:
Merge branch 'main' into aho-corasick

stephentoub commented:
@teo-tsirpanis, just FYI as to status, @MihaZupan has been looking at a vectorized solution for the same set of target scenarios.  If that goes as well as we hope, I expect the likely outcome is we'll use some of your PR's logic for extracting the relevant portions of the regex node tree and combine that with Miha's work rather than the Aho-Corasick implementation in this PR.

teo-tsirpanis commented:
Great news! Let me know when such PR gets opened, I can take a look at it.

MihaZupan commented:
Here's the code for it so far: https://github.com/dotnet/runtime/compare/main...MihaZupan:runtime:teddy
It's `IndexOfAnyValues<string>` and a `static int IndexOfAny(this ROS<char>, IndexOfAnyValues<string>)` extension that is then backed by combinations of Teddy/Rabin-Karp/Aho-Corasick.

teo-tsirpanis commented:
I'm not going to reopen this, looking forward to performance improvements in regexes through ~~`IndexOfAnyValues`~~ `SearchValues`. ✌🏻
